network/src/sync/blocks/aggro.rs

```diff
@@ -0,0 +1,202 @@
+                                let received_hashes = received_hashes.read().await;
+                                let hashes = block_locator_hashes.read().await;
+                                hashes
+                                    .hashes
+                                    .iter()
+                                    .filter(|x| !received_hashes.contains(&x.0[..]))
+                                    .cloned()
+                                    .chain(
+                                        hashes_trimmed
+                                            .into_iter()
+                                            .zip(early_block_states.iter())
+                                            .filter(|(_, status)| {
+                                                matches!(status, snarkos_storage::BlockStatus::Unknown)
+                                            })
+                                            .map(|(hash, _)| BlockHeaderHash(hash.bytes().unwrap())),
+                                    )
+                                    .collect()
+                            };
+                            if blocks.is_empty() {
+                                return;
+                            }
+                            debug!("requesting {} sync blocks from {}", blocks.len(), peer);
+
+                            if let Some(peer) = node.peer_book.get_peer_handle(peer) {
+                                let request: Vec<BlockHeaderHash> = blocks.into_iter().collect();
+                                peer.expecting_sync_blocks(request.len() as u32).await;
+                                peer.send_payload(Payload::GetBlocks(request), None).await;
+                                peer.send_payload(
+                                    Payload::GetSync(vec![BlockHeaderHash(last_hash.bytes().unwrap())]),
+                                    None,
+                                )
+                                .await;
+                            }
+                        });
+                    }
+                    SyncInbound::Block(peer, block, peer_height) => {
+                        let node = node.clone();
+                        let received_hashes = received_hashes.clone();
+                        tokio::spawn(async move {
+                            received_hashes.write().await.push(&block[..]);
+                            match node.process_received_block(peer, block, peer_height, false).await {
+                                Err(e) => warn!("failed to process received block from {}: {:?}", peer, e),
+                                Ok(()) => (),
+                            }
+                        });
+                    }
+                }
+                false
+            })
+            .await;
+
+        let sync_addresses = sync_nodes.iter().map(|x| x.address).collect::<Vec<_>>();
+
+        self.base.cancel_outstanding_syncs(&sync_addresses[..]).await;
```

Comment:
It wasn't immediately obvious to me why this needs to be called at the end of every sync, it would be nice to comment on this here? (It may also not be necessary if state is tracked on the sync layer.)

Reply:
it notifies each peer to reset tracking state. Agreed.